addons/crm/static/src/js/forecast/forecast_model_extension.js

```diff
@@ -0,0 +1,117 @@
+/** @odoo-module alias=crm.ForecastModelExtension **/
+import ActionModel from 'web.ActionModel';
+
+/**
+ * File intended for `web` (could be used in all kind of forecast configurations)
+ * 
+ * This file contains the logic behind a special "Forecast" filter.
+ * Any such filter should set the context key {forecast_filter: 1}
+ * Another context key must also be set for the view using this model extension:
+ * {forecast_field: "date_field_name"}, which represents the date/time field on which
+ * the forecast should be applied
+ * 
+ * The main purpose is to be able to modify the domain depending on the groupby granularity
+ * when the field is a `date/time` field and is the `forecast_field`. The domain should filter
+ * records from the period (granularity) containing "now", or those where the forecast_field
+ * is not set.
+ * example:
+ *  today:          2021-04-21
+ *  granularity:    month
+ *  field name:     date_field
+ *  -> the domain would be: ['&', ['date_field', '=', false], ['date_field', '>=', '2021-04-01']]
+ */
+class ForecastModelExtension extends ActionModel.Extension {
+
+    //---------------------------------------------------------------------
+    // Public
+    //---------------------------------------------------------------------
+
+    /**
+     * @override
+     * @returns {any}
+     */
+    get(property) {
```

Comment:
Strange this does not call super.

And while we are at it: did not find where it is used ? seems we call getDomain on Service, but did not find call on model. Is it automatically done behind the scene ?

Reply:
This is the abstract function that we override:
https://github.com/odoo/odoo/blob/4aa89f146e2d6c03df5803dbb196caf771269b35/addons/web/static/src/js/model.js#L61-L71
The process to load the extension and get the modified domain starts here:
https://github.com/odoo/odoo/blob/4aa89f146e2d6c03df5803dbb196caf771269b35/addons/web/static/src/js/views/abstract_view.js#L310-L311
`searchModel.load` -> ... -> `model._loadExtensions`
At this point, the extension is ready to provide the modified domain, which is requested by the following:
`searchModel.get("query")` -> ... -> `actionModel._getQuery` -> ... -> `model.__get` ->
https://github.com/odoo/odoo/blob/4aa89f146e2d6c03df5803dbb196caf771269b35/addons/web/static/src/js/model.js#L399

Reply:
Thanks for your explanations.

Shouldn't we call super() quand même, even if original code does nothing ? I mean as this is a switch/case, would makes sense to enable overrides.